graph2tac/loader/data_server.py

```diff
@@ -171,21 +172,95 @@ def cluster_to_graph(self, cluster):
             ], dtype = self._def_name_dtype)
         )
 
-    
+TRAIN = 0
+VALID = 1
+HOLDOUT = 2 # maybe unnecessary
+
+class Splitter:
+    def lemma(self, d : Definition) -> int:
+        raise Exception("Not implemented")
+    def definition_cluster(self, d : list[Definition]) -> int:
+        raise Exception("Not implemented")
+    def assign_data_server(self, data_server):
+        pass
+
+class SplitByHash(Splitter):
+    def __init__(self, proportions : list[int], random_seed : int):
+        self.proportions = proportions
+        self.random_seed = random_seed
+    def lemma(self, d : Definition) -> int:
+        # to make it identical to vasily's loader
+        ident_64 = np.array(d.node.identity, dtype = np.uint64).item()
+        return get_split_label(ident_64, self.proportions, self.random_seed)
+    def definition_cluster(self, d : list[Definition]) -> int:
+        return 0
+
+class SplitByFilePrefix(Splitter):
+    def __init__(self, prefixes_per_label : list[list[str]]):
+        self.prefixes_per_label = [
+            (label+1, prefixes)
+            for label, prefixes in enumerate(prefixes_per_label)
+        ]
+    def lemma(self, d : Definition):
+        return self.graphid_to_label[d.node.graph]
+    def definition_cluster(self, d : list[Definition]):
+        return self.graphid_to_label[d[0].node.graph]
+    def assign_data_server(self, data_server):
+
+        # conversion from graph indices to filenames
+        graphid_to_fname = {
+            file_data.graph : str(fname)
+            for fname, file_data in data_server._data.items()
+        }
+        assert set(graphid_to_fname.keys()) == set(range(len(data_server._data)))
+        graphid_to_fname = [
+            graphid_to_fname[i]
+            for i in range(len(data_server._data))
+        ]
```

Comment:
Is all of this really needed? As far as I can tell, you never actually look something up in these maps. You only use the domain of the map.

Reply:
It is a way to ensure that every graphid `0..len(data_server._data)-1` occurs there exactly once.

Reply:
You mean that you want to check that there is a bijection between filenames and graphid's? Okay, fine I guess, although if we want to test this, it might be better to do that centrally in pytact. (Although it seems slightly redundant, because in pytact essentially the opposite computation is happening to the one here...)

Reply:
I have to do the inverse, and
```
        graphid_to_fname = {
            file_data.graph : str(fname)
            for fname, file_data in data_server._data.items()
        }
```
is in my view the most intuitive way of writing an inverse. Yes, I could keep it in a dictionary, and not convert to a list (but lists have some further advantages).

Reply:
Okay, this feels a little convoluted, but I see your point.